### PR TITLE
2510 UAC by SMS 

### PR DESCRIPTION
**Pull request by AdamHawtin**
# Motivation and Context
We have a new API service for SMS fulfilments, the case processor needs to log these and link their UACs. 

# What has changed
- Add new receiver for (enriched) SMS fulfilment events
- Tests
- Add new SMS template entities
- Update README

# How to test?
Checkout and build this branch locally, then the other service branches linked on the ticket.
Run the acceptance tests branch.

# Links
https://trello.com/c/uQcwIQcQ/2510-prototype-send-a-uac-by-sms-by-api-13

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -45,6 +58,33 @@ Example message:
 }
 ```
 
+### Enriched SMS Fulfilment
+Topic: `rm-internal-sms-fulfilment_case-processor`
+
+Example message:
+```json
+{
+  "header": {
```
Comment (NickGrantONS) [inline README.md:67]: is it worth versioning?

File: src/main/java/uk/gov/ons/ssdc/caseprocessor/messaging/SmsFulfilmentReceiver.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+  public SmsFulfilmentReceiver(
+      UacService uacService, CaseService caseService, EventLogger eventLogger) {
+    this.uacService = uacService;
+    this.caseService = caseService;
+    this.eventLogger = eventLogger;
+  }
+
+  @Transactional
+  @ServiceActivator(inputChannel = "smsFulfilmentInputChannel")
```
Comment (NickGrantONS) [inline src/main/java/uk/gov/ons/ssdc/caseprocessor/messaging/SmsFulfilmentReceiver.java:37]:
```suggestion
  @ServiceActivator(inputChannel = "smsFulfilmentInputChannel", adviceChain = "retryAdvice")
```

File: src/main/java/uk/gov/ons/ssdc/caseprocessor/model/dto/EventTypeDTO.java [hunk 3]
```diff
@@ -0,0 +1 @@
+
```
Comment (NickGrantONS) [inline src/main/java/uk/gov/ons/ssdc/caseprocessor/model/dto/EventTypeDTO.java:1]: This seems to be blank... class should probably be deleted?

File: src/main/java/uk/gov/ons/ssdc/caseprocessor/model/entity/SmsTemplate.java [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+@Entity
+public class SmsTemplate {
+  @Id private String packCode;
+
+  @Type(type = "jsonb")
+  @Column(columnDefinition = "jsonb", nullable = false)
+  private String[] template;
+
+  @Column(nullable = false)
+  private UUID notifyId;
```
Comment (NickGrantONS) [inline src/main/java/uk/gov/ons/ssdc/caseprocessor/model/entity/SmsTemplate.java:20]: what is the `notifyId` for?

Comment (NickGrantONS) [reply src/main/java/uk/gov/ons/ssdc/caseprocessor/model/entity/SmsTemplate.java:20]: is it the notify template ID? If so, it should probably have the word "template" in the name

Comment (NickGrantONS) [inline src/main/java/uk/gov/ons/ssdc/caseprocessor/model/entity/SmsTemplate.java:20]:
```suggestion
  private UUID notifyTemplateId;
```

File: src/main/java/uk/gov/ons/ssdc/caseprocessor/service/UacService.java [hunk 5]
```diff
@@ -66,4 +68,16 @@ public UacQidLink findByQid(String questionnaireId) {
...
   }
+
+  public void createLinkAndEmitNewUacQid(Case caze, String uac, String qid) {
+    OffsetDateTime now = OffsetDateTime.now();
+    UacQidLink uacQidLink = new UacQidLink();
+    uacQidLink.setId(UUID.randomUUID());
+    uacQidLink.setUac(uac);
+    uacQidLink.setQid(qid);
+    uacQidLink.setCaze(caze);
+    uacQidLink.setCreatedAt(now);
```
Comment (NickGrantONS) [inline src/main/java/uk/gov/ons/ssdc/caseprocessor/service/UacService.java:79]: This should be done automatically by Hibernate

File: src/main/java/uk/gov/ons/ssdc/caseprocessor/service/UacService.java [hunk 6]
```diff
@@ -66,4 +68,16 @@ public UacQidLink findByQid(String questionnaireId) {
...
+    uacQidLink.setLastUpdatedAt(now);
```
Comment (NickGrantONS) [inline src/main/java/uk/gov/ons/ssdc/caseprocessor/service/UacService.java:80]: This should be done automatically by Hibernate

File: src/main/java/uk/gov/ons/ssdc/caseprocessor/service/UacService.java [hunk 7]
```diff
@@ -66,4 +68,16 @@ public UacQidLink findByQid(String questionnaireId) {
   public boolean existsByQid(String qid) {
     return uacQidLinkRepository.existsByQid(qid);
   }
+
+  public void createLinkAndEmitNewUacQid(Case caze, String uac, String qid) {
+    OffsetDateTime now = OffsetDateTime.now();
```
Comment (NickGrantONS) [inline src/main/java/uk/gov/ons/ssdc/caseprocessor/service/UacService.java:73]: The created/updated timestamps should be done automatically by Hibernate

File: src/test/java/uk/gov/ons/ssdc/caseprocessor/messaging/TelephoneCaptureReceiverIT.java [hunk 8]
```diff
@@ -36,10 +37,9 @@
 @ActiveProfiles("test")
 @SpringBootTest
 @ExtendWith(SpringExtension.class)
-public class TelephoneCaptureReceiverIT {
+class TelephoneCaptureReceiverIT {
```
Comment (NickGrantONS) [inline src/test/java/uk/gov/ons/ssdc/caseprocessor/messaging/TelephoneCaptureReceiverIT.java:40]: class should remain public

File: src/test/java/uk/gov/ons/ssdc/caseprocessor/messaging/TelephoneCaptureReceiverIT.java [hunk 9]
```diff
@@ -58,7 +58,7 @@ public void setUp() {
   }
 
   @Test
-  public void testTelephoneCapture() throws Exception {
+  void testSmsFulfilment() throws Exception {
```
Comment (NickGrantONS) [inline src/test/java/uk/gov/ons/ssdc/caseprocessor/messaging/TelephoneCaptureReceiverIT.java:61]: test method should remain public

File: src/test/java/uk/gov/ons/ssdc/caseprocessor/messaging/TelephoneCaptureReceiverTest.java [hunk 10]
```diff
@@ -28,7 +26,7 @@
 import uk.gov.ons.ssdc.caseprocessor.service.UacService;
 
 @ExtendWith(MockitoExtension.class)
-public class TelephoneCaptureReceiverTest {
+class TelephoneCaptureReceiverTest {
```
Comment (NickGrantONS) [inline src/test/java/uk/gov/ons/ssdc/caseprocessor/messaging/TelephoneCaptureReceiverTest.java:29]: class should remain public

File: src/test/java/uk/gov/ons/ssdc/caseprocessor/messaging/TelephoneCaptureReceiverTest.java [hunk 11]
```diff
@@ -43,7 +41,7 @@
   private static final String TELEPHONE_CAPTURE_DESCRIPTION = "Telephone capture request received";
 
   @Test
-  public void testReceiveMessageHappyPath() {
+  void testReceiveMessageHappyPath() {
```
Comment (NickGrantONS) [inline src/test/java/uk/gov/ons/ssdc/caseprocessor/messaging/TelephoneCaptureReceiverTest.java:44]: test method should remain public

File: src/test/java/uk/gov/ons/ssdc/caseprocessor/messaging/TelephoneCaptureReceiverTest.java [hunk 12]
```diff
@@ -77,7 +69,7 @@ public void testReceiveMessageHappyPath() {
   }
 
   @Test
-  public void testReceiveMessageQidAlreadyLinkedToCorrectCase() {
+  void testReceiveMessageQidAlreadyLinkedToCorrectCase() {
```
Comment (NickGrantONS) [inline src/test/java/uk/gov/ons/ssdc/caseprocessor/messaging/TelephoneCaptureReceiverTest.java:72]: test method should remain public

File: src/test/java/uk/gov/ons/ssdc/caseprocessor/service/UacServiceTest.java [hunk 13]
```diff
@@ -29,7 +30,7 @@
   @InjectMocks UacService underTest;
 
   @Test
-  public void saveAndEmitUacUpdatedEvent() {
+  void saveAndEmitUacUpdatedEvent() {
```
Comment (NickGrantONS) [inline src/test/java/uk/gov/ons/ssdc/caseprocessor/service/UacServiceTest.java:33]: test method should remain public... should also be prefixed `test` but that's another story

File: src/test/java/uk/gov/ons/ssdc/caseprocessor/service/UacServiceTest.java [hunk 14]
```diff
@@ -52,7 +53,7 @@ public void saveAndEmitUacUpdatedEvent() {
   }
 
   @Test
-  public void findByQid() {
+  void findByQid() {
```
Comment (NickGrantONS) [inline src/test/java/uk/gov/ons/ssdc/caseprocessor/service/UacServiceTest.java:56]: test method should remain public

File: src/test/java/uk/gov/ons/ssdc/caseprocessor/service/UacServiceTest.java [hunk 15]
```diff
@@ -66,7 +67,7 @@ public void findByQid() {
   }
 
   @Test
-  public void findByQidFails() {
+  void findByQidFails() {
```
Comment (NickGrantONS) [inline src/test/java/uk/gov/ons/ssdc/caseprocessor/service/UacServiceTest.java:70]: test method should remain public

File: src/test/java/uk/gov/ons/ssdc/caseprocessor/service/UacServiceTest.java [hunk 16]
```diff
@@ -76,9 +77,43 @@ public void findByQidFails() {
   }
 
   @Test
-  public void existsByQid() {
+  void existsByQid() {
```
Comment (NickGrantONS) [inline src/test/java/uk/gov/ons/ssdc/caseprocessor/service/UacServiceTest.java:80]: test method should remain public

File: src/test/java/uk/gov/ons/ssdc/caseprocessor/service/UacServiceTest.java [hunk 17]
```diff
@@ -76,9 +77,43 @@ public void findByQidFails() {
...
     String testQid = "12345";
     underTest.existsByQid(testQid);
     verify(uacQidLinkRepository).existsByQid(testQid);
   }
+
+  @Test
+  void createLinkAndEmitNewUacQid() {
```
Comment (NickGrantONS) [inline src/test/java/uk/gov/ons/ssdc/caseprocessor/service/UacServiceTest.java:87]: test method should remain public

File: src/main/resources/application.yml [hunk 18]
```diff
@@ -44,6 +44,7 @@ queueconfig:
   case-update-topic: event_case-update
   uac-update-topic: event_uac-update
   deactivate-uac-topic: event_deactivate-uac
+  sms-fulfilment-subscription: rm-internal-sms-fulfilment_case-processor
```
Comment (NickGrantONS) [inline src/main/resources/application.yml:47]: 🎨 I would probably keep the subscriptions grouped together, personally... although I see that it hasn't been followed consistently (probably by me) anyway

**Review by NickGrantONS — CHANGES_REQUESTED** [review]
Great work. Left a few comments

File: src/test/java/uk/gov/ons/ssdc/caseprocessor/messaging/TelephoneCaptureReceiverIT.java (hunk 8 above)
Comment (AdamHawtin) [reply src/test/java/uk/gov/ons/ssdc/caseprocessor/messaging/TelephoneCaptureReceiverIT.java:40]: The linter was complaining at me to scrap them, don't mind changing the linting rules instead if you disagree.

Comment (NickGrantONS) [reply src/test/java/uk/gov/ons/ssdc/caseprocessor/messaging/TelephoneCaptureReceiverIT.java:40]: hmmm, so it does seem that JUnit5 style is to leave them `protected`

Comment (NickGrantONS) [reply src/test/java/uk/gov/ons/ssdc/caseprocessor/messaging/TelephoneCaptureReceiverIT.java:40]: tested in IDE and it seems fine. Guess it was JUnit 4 which was more fussy

**Review by NickGrantONS — APPROVED** [review]
ATs passing. Great work

**Review by liamtoozer — APPROVED** [review]
LGTM 👍  Tested locally

**Pull request merged**